src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,353 @@
+          && lineNumber - 1 == 0
+          && partitionStartByteRange != 0) {
+        InputStreamReader startStreamReader =
+            new InputStreamReader(
+                source.getInputStream(partitionStartByteRange - 1, partitionEndByteRange));
+        char currentChar = (char) startStreamReader.read();
+        if (currentChar != '\n' && currentChar != '\r') {
+          discardedRecord = true;
+        }
+      }
+      // This covers the case where the last row matches the schema's number of columns. We need to ensure that we read
+      // the entire row completely
+      else if (record.size() == schema.numColumns() && !iterator.hasNext() && !isLastWorker) {
+        long movingEndByte = partitionEndByteRange;
+        long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+        boolean newLineFound = false;
+        while (!newLineFound) {
+          movingEndByte += MyriaConstants.BYTE_OVERLAP_PARALLEL_INGEST;
+          // Create a stream to look for the new line
+          InputStream trailingEndInputStream =
+              source.getInputStream(partitionEndByteRange, movingEndByte);
+          InputStreamReader startStreamReader = new InputStreamReader(trailingEndInputStream);
+          int dataChar = startStreamReader.read();
+          while (dataChar != -1) {
+            char currentChar = (char) dataChar;
+            if (currentChar == '\n' || currentChar == '\r') {
+              newLineFound = true;
+              // Re-initialize the parser with the last row only
+              InputStream beginningOfRecord =
+                  source.getInputStream(bytePositionAtBeginningOfRecord, partitionEndByteRange);
+              InputStream concatenateEndOfRecord =
+                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
+              partitionInputStream =
+                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
+              parser =
+                  new CSVParser(
+                      new BufferedReader(new InputStreamReader(partitionInputStream)),
+                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
+                      bytePositionAtBeginningOfRecord,
+                      0);
+              iterator = parser.iterator();
+              onLastRow = true;
+              record = iterator.next();
+              break;
+            }
+            dataChar = startStreamReader.read();
+          }
+          startStreamReader.close();
+        }
+      }
+      // This is a partial fragment, columns are missing and we need to read more bytes
+      if (record.size() < schema.numColumns()) {
+        if (lineNumber - 1 != 0 && !isLastWorker) {
+          onLastRow = true;
+          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+          if (!fixingStartByte) {
+            fixingStartByte = true;
+            partitionStartByteRange += bytePositionAtBeginningOfRecord;
```

Comment:
This approach isn't possible since we can't mix byte and character offsets. In any case, I don't think it's necessary: we can just keep the original `InputStream` for the partition around and keep concatenating exponentially larger `InputStream`s (the size of `byteOverlap`) to the end of a `SequenceInputStream`. Then just pass `charPositionAtBeginningOfRecord` to the `CSVParser` constructor.